File: graph/scc/scc.py
import sys
input = sys.stdin.readline

def scc(node):
    global ids, scc_cnt
    sstack = [(node, 0, -1)]
    while sstack:
        node, idx, ret = sstack.pop()
        if ids_arr[node] == 0 and ret == -1 and idx == 0:
            ids_arr[node], parents[node] = ids, ids
            ids += 1
            visited[node] = True
            stack.append(node)
        if ret != -1:
            child = graph[node][ret]
            parents[node] = min(parents[child], parents[node])
            idx = ret+1 
        while idx < len(graph[node]):
            now = graph[node][idx]
            if ids_arr[now] == 0:
                sstack.append((node, idx, idx))
                sstack.append((now, 0, -1))
                break
            elif visited[now]: parents[node] = min(parents[now], parents[node])
            idx += 1
        else:
            w = -1
            if parents[node] == ids_arr[node]:
                while w != node:
                    w = stack.pop()
                    scc_idx[w] = scc_cnt
                    visited[w] = False
                scc_cnt += 1

v, e = map(int, input().split())
graph = [[] for _ in range((v+1))]
for _ in range(e):
    a, b = map(int, input().split())
    graph[a].append(b)
stack = []
visited = [False]*(v+1)
ids, ids_arr = 1, [0]*(v+1)
parents, scc_idx, scc_cnt = [0]*(v+1), [-1]*(v+1), 0

for i in range(1, v+1):
    if ids_arr[i] == 0: scc(i)

indegree = [0 for _ in range(scc_cnt)]
for i in range(1, v+1):
    for j in graph[i]:
        if scc_idx[i] != scc_idx[j]:
            indegree[scc_idx[j]] += 1

File: graph/scc/scc.cpp
#include <bits/stdc++.h>
#define fastio cin.tie(0), cout.tie(0), ios_base::sync_with_stdio(false)
#define MAX 1000001
typedef long long ll;
using namespace std;

vector<ll> stacks;
vector<vector<ll>> graph(MAX);
vector<vector<ll>> ans(MAX);
bool visited[MAX];
ll ids_arr[MAX], parents[MAX], scc_idx[MAX], indegree[MAX];
ll ids = 1, scc_cnt = 0;
ll n, m, a, b;


void scc(ll node){
    ids_arr[node] = ids; parents[node] = ids;
    ids++; visited[node] = true;
    stacks.emplace_back(node);
    
    for (auto now: graph[node]){
        if (ids_arr[now] == 0){
            scc(now);
            parents[node] = min(parents[now], parents[node]);
        }
        else if (visited[now])
            parents[node] = min(parents[now], parents[node]);
    }
    
    ll w = -1;
    if (parents[node] == ids_arr[node]){
        while (w != node){
            w = stacks.back();
            stacks.pop_back();
            scc_idx[w] = scc_cnt;
            visited[w] = false;
        }
        scc_cnt++;
    }
}

int main() {
    fastio;
    cin >> n >> m;
    for(ll i=0; i<m; i++){
        cin >> a >> b;
        graph[a].emplace_back(b);
    }
    for(ll i=1; i<=n; i++)
        if (ids_arr[i] == 0) scc(i);
    
    for(ll i=1; i<=n; i++){
        ans[scc_idx[i]].push_back(i);
        for (auto j: graph[i]){
            if (scc_idx[i] != scc_idx[j]) indegree[scc_idx[j]]++;
        }
    }
}
